server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -189,7 +191,17 @@ public static <Response extends ReplicationResponse & WriteResponse> ActionListe
     }
 
     @Override
-    protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<BulkResponse> listener) {
+    protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<BulkResponse> bulkResponseListener) {
+        final ActionListener<BulkResponse> listener = bulkResponseListener.delegateFailure((l, r) -> {
+            if (DiscoveryNode.isStateless(clusterService.getSettings())
+                && bulkRequest.getRefreshPolicy() != WriteRequest.RefreshPolicy.NONE) {
+                // As a work-around to support `?refresh`, explicitly make a call to the Refresh API.
+                // TODO: Replace with a less hacky approach.
+                client.admin().indices().prepareRefresh().execute(l.map(ignored -> r));
+            } else {
+                l.onResponse(r);
+            }
+        });
```

Comment:
I know this is what I suggested in Slack, but now I look at it I think it'd be better to avoid the extra wrapping when unnecessary:

```java
 // As a work-around to support `?refresh`, explicitly make a call to the Refresh API.
        // TODO: Replace with a less hacky approach.
        final ActionListener<BulkResponse> listener = DiscoveryNode.isStateless(clusterService.getSettings())
            && bulkRequest.getRefreshPolicy() != WriteRequest.RefreshPolicy.NONE
                ? bulkResponseListener.delegateFailure((l, r) -> client.admin().indices().prepareRefresh().execute(l.map(ignored -> r)))
                : bulkResponseListener;
```

Reply:
Also both listeners are bulk-response listeners, I prefer the name `outerListener`